core/trino-main/src/main/java/io/trino/operator/aggregation/OrderedWindowAccumulator.java

```diff
@@ -0,0 +1,156 @@
+        requireNonNull(sortOrders, "sortOrders is null");
+        requireNonNull(sortKeysArguments, "sortChannels is null");
+        checkArgument(sortOrders.size() == sortKeysArguments.size(), "sortOrders and sortChannels must have the same size");
+        sortKeysArguments.forEach(argument -> {
+            checkArgument(
+                    argument < argumentChannels.size(),
+                    "invalid argument %s referenced; total number of arguments is %s", argument, argumentChannels.size());
+        });
+        this.sortOrders = ImmutableList.copyOf(sortOrders);
+        this.sortKeysArguments = ImmutableList.copyOf(sortKeysArguments);
+
+        this.delegate = requireNonNull(delegate, "delegate is null");
+        this.initialDelegate = delegate.copy();
+
+        this.pageBuilder = new PageBuilder(argumentTypes);
+    }
+
+    @Override
+    public long getEstimatedSize()
+    {
+        return delegate.getEstimatedSize() + initialDelegate.getEstimatedSize() + pagesIndex.getEstimatedSize().toBytes() + pageBuilder.getRetainedSizeInBytes();
+    }
+
+    @Override
+    public WindowAccumulator copy()
+    {
+        PagesIndex pagesIndexCopy = pagesIndexFactory.newPagesIndex(argumentTypes, pagesIndex.getPositionCount());
+        pagesIndex.getPages().forEachRemaining(pagesIndexCopy::addPage);
+        return new OrderedWindowAccumulator(pagesIndexFactory, pagesIndexCopy, delegate.copy(), argumentTypes, argumentChannels, sortKeysArguments, sortOrders);
+    }
+
+    @Override
+    public void addInput(WindowIndex index, int startPosition, int endPosition)
+    {
+        if (pagesIndexSorted) {
+            pagesIndexSorted = false;
+            // operate on delegate as of start
+            // nicer would be to add reset() method to WindowAccumulator but it requires reset method in each AccumulatorState class
+            delegate = initialDelegate.copy();
+        }
+        // index is remapped so just go from 0 to argumentChannels.size()
+        for (int position = startPosition; position <= endPosition; position++) {
+            if (pageBuilder.isFull()) {
+                indexCurrentPage();
+            }
+            for (int channel = 0; channel < argumentChannels.size(); channel++) {
+                ValueBlock value = index.getSingleValueBlock(channel, position).getSingleValueBlock(0);
+                pageBuilder.getBlockBuilder(channel).append(value, 0);
+            }
+            pageBuilder.declarePosition();
+        }
+    }
+
+    private void indexCurrentPage()
+    {
+        pagesIndex.addPage(pageBuilder.build());
+        pageBuilder.reset();
+    }
+
+    @Override
+    public void output(BlockBuilder blockBuilder)
+    {
+        if (!pagesIndexSorted) {
+            if (!pageBuilder.isEmpty()) {
+                indexCurrentPage();
+            }
+            int positionCount = pagesIndex.getPositionCount();
+            if (positionCount == 0) {
+                return;
```

Comment:
For an empty frame it should append a scalar result. `delegate.output(blockBuilder)` should do the job. Please add a test.

Reply:
good catch.